lib/path-design.md

```diff
@@ -0,0 +1,177 @@
+- :heavy_minus_sign: `..` are a part of paths, a path library should therefore support it
+  - :heavy_plus_sign: If we can prove that all such use cases are better done e.g. with runtime tools, the library not supporting it can nudge people towards that
+    - :heavy_minus_sign: Can we prove that though?
+- :heavy_minus_sign: We could allow ".." just in the beginning
+  - :heavy_plus_sign: Then we'd have to throw an error for doing `append /some/path "../foo"`, making it non-composable
+  - :heavy_plus_sign: The same is for returning paths with `..`: `relativeTo /foo /bar => "../foo"` would produce a non-composable path
+- :heavy_plus_sign: We argue that `..` is not needed at the Nix evaluation level, since we'd always start evaluation from the project root and don't go up from there
+  - :heavy_plus_sign: And `..` is supported in Nix paths, turning them into absolute paths
+    - :heavy_minus_sign: This is ambiguous with symlinks though
+- :heavy_plus_sign: If you need `..` for building or runtime, you can use build/run-time tooling to create those (e.g. `realpath` with `--relative-to`), or use absolute paths instead.
+  This also gives you the ability to correctly handle symlinks
+
+</details>
+
+### Trailing slashes
+[trailing-slashes]: #trailing-slashes
+
+Context: Relative paths can contain trailing slashes, like `foo/`, indicating that the path points to a directory and not a file
+
+Decision: All functions remove trailing slashes in their results
+
+<details>
+<summary>Arguments</summary>
+
+- :heavy_plus_sign: It enables the law that if `normalise p == normalise q` then `$(stat p) == $(stat q)`.
+- Comparison to other frameworks to figure out the least surprising behavior:
+  - :heavy_plus_sign: Nix itself doesn't preserve trailing newlines when parsing and appending its paths
+  - :heavy_minus_sign: [Rust's std::path](https://doc.rust-lang.org/std/path/index.html) does preserve them during [construction](https://doc.rust-lang.org/std/path/struct.Path.html#method.new)
+    - :heavy_plus_sign: Doesn't preserve them when returning individual [components](https://doc.rust-lang.org/std/path/struct.Path.html#method.components)
+    - :heavy_plus_sign: Doesn't preserve them when [canonicalizing](https://doc.rust-lang.org/std/path/struct.Path.html#method.canonicalize)
+  - :heavy_plus_sign: [Python 3's pathlib](https://docs.python.org/3/library/pathlib.html#module-pathlib) doesn't preserve them during [construction](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath)
+    - Notably it represents the individual components as a list internally
+  - :heavy_minus_sign: [Haskell's filepath](https://hackage.haskell.org/package/filepath-1.4.100.0) has [explicit support](https://hackage.haskell.org/package/filepath-1.4.100.0/docs/System-FilePath.html#g:6) for handling trailing slashes
```

Comment:
I feel like we can leave that one out, since it doesn't treat paths as strings, so there's no normalisation necessary. The normalisation is the type structure itself.